api-put-object.go

```diff
@@ -143,164 +141,71 @@ func (a completedParts) Less(i, j int) bool { return a[i].PartNumber < a[j].Part
-	// Return error if its not io.EOF.
-	if err != nil && err != io.EOF {
-		return 0, err
-	}
+// PutObjectWithMetadata using AWS streaming signature V4
+func (c Client) PutObjectWithMetadata(bucketName, objectName string, reader io.Reader, metadata map[string][]string, progress io.Reader) (n int64, err error) {
+	return c.PutObjectWithProgress(bucketName, objectName, reader, metadata, progress)
+}
 
-	// Seek back to beginning of the temporary file.
-	if _, err = tmpFile.Seek(0, 0); err != nil {
-		return 0, err
-	}
-	reader = tmpFile
+// PutObjectWithProgress using AWS streaming signature V4
+func (c Client) PutObjectWithProgress(bucketName, objectName string, reader io.Reader, metadata map[string][]string, progress io.Reader) (n int64, err error) {
+	// Size of the object.
+	var size int64
 
-	// Execute put object.
-	st, err := c.putObjectDo(bucketName, objectName, reader, hashSums["md5"], hashSums["sha256"], size, metaData)
+	// Get reader size.
+	size, err = getReaderSize(reader)
 	if err != nil {
 		return 0, err
 	}
-	if st.Size != size {
-		return 0, ErrUnexpectedEOF(st.Size, size, bucketName, objectName)
-	}
-	// Progress the reader to the size if putObjectDo is successful.
-	if progress != nil {
-		if _, err = io.CopyN(ioutil.Discard, progress, size); err != nil {
-			return size, err
-		}
-	}
-	return size, nil
+	return c.putObjectCommon(bucketName, objectName, reader, size, metadata, progress)
 }
 
-// putObjectDo - executes the put object http operation.
-// NOTE: You must have WRITE permissions on a bucket to add an object to it.
-func (c Client) putObjectDo(bucketName, objectName string, reader io.Reader, md5Sum []byte, sha256Sum []byte, size int64, metaData map[string][]string) (ObjectInfo, error) {
-	// Input validation.
-	if err := s3utils.CheckValidBucketName(bucketName); err != nil {
-		return ObjectInfo{}, err
+func (c Client) putObjectCommon(bucketName, objectName string, reader io.Reader, size int64, metadata map[string][]string, progress io.Reader) (n int64, err error) {
+	// Check for largest object size allowed.
+	if size > int64(maxMultipartPutObjectSize) {
+		return 0, ErrEntityTooLarge(size, maxMultipartPutObjectSize, bucketName, objectName)
 	}
-	if err := s3utils.CheckValidObjectName(objectName); err != nil {
-		return ObjectInfo{}, err
-	}
-
-	// Set headers.
-	customHeader := make(http.Header)
 
-	// Set metadata to headers
-	for k, v := range metaData {
-		if len(v) > 0 {
-			customHeader.Set(k, v[0])
-		}
+	// NOTE: Streaming signature is not supported by GCS.
+	if s3utils.IsGoogleEndpoint(c.endpointURL) {
+		// Do not compute MD5 for Google Cloud Storage.
+		return c.putObjectNoChecksum(bucketName, objectName, reader, size, metadata, progress)
 	}
 
-	// If Content-Type is not provided, set the default application/octet-stream one
-	if v, ok := metaData["Content-Type"]; !ok || len(v) == 0 {
-		customHeader.Set("Content-Type", "application/octet-stream")
+	if c.overrideSignerType.IsV2() {
```

Comment:
hi, why does V2 singer can't use streaming upload?

Reply:
Streaming standard is only signature v4

Reply:
thank you, It seems that I could steaming upload object with signature V2. is there any problems if we remove this check?

Reply:
@harshavardhana "Streaming standard is only signature v4" do you mean this [Transferring Payload in Multiple Chunks](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming.html).  but minio-go's [executeMethod](https://github.com/minio/minio-go/blob/46b02d3b4e8c63804a5d4d4d14a562fd71c66730/api.go#L524) do not use this chunk upload whether it's V2 or V4 signature.

Reply:
of course, we do @Liuchang0812 

```
   switch {
        case signerType.IsV2():
                // Add signature version '2' authorization header.
                req = s3signer.SignV2(*req, accessKeyID, secretAccessKey, isVirtualHost)
        case metadata.objectName != "" && method == "PUT" && metadata.customHeader.Get("X-Amz-Copy-Source") == "" && !c.secure:
                // Streaming signature is used by default for a PUT object request. Additionally we also
                // look if the initialized client is secure, if yes then we don't need to perform
                // streaming signature.
                req = s3signer.StreamingSignV4(req, accessKeyID,
                        secretAccessKey, sessionToken, location, metadata.contentLength, time.Now().UTC())
        default:
                // Set sha256 sum for signature calculation only with signature version '4'.
                shaHeader := unsignedPayload
                if metadata.contentSHA256Hex != "" {
                        shaHeader = metadata.contentSHA256Hex
                }
                req.Header.Set("X-Amz-Content-Sha256", shaHeader)

                // Add signature version '4' authorization header.
                req = s3signer.SignV4(*req, accessKeyID, secretAccessKey, sessionToken, location)
        }
```

Reply:
@hackintoshrao thanks for your help. `s3signer.SignV4` does not sign http body as [transferring payload in multiple chunks](s3signer.SignV4). please let me know if i missing anything.

what I meet is minio-go consumes nearly 600MB memory when I upload a object with V2 sign. because minio-go allocate 64MB memory [here](https://github.com/minio/minio-go/blob/master/api-put-object-multipart.go#L100).

it seems that the diff between `putObjectMultipartNoStream` and `putObjectMultipartStreamNoChecksum` is non-stream api has to read file content to memory and calculate MD5/SHA.